### ISSUE TITLE
"find" erroring after network data migration to NetApp FAS2620.

### ISSUE THREAD

**Issue by KelSolaar** [git-for-windows/git#1497]
- [X] I was not able to find an [open](https://github.com/git-for-windows/git/issues?q=is%3Aopen) or [closed](https://github.com/git-for-windows/git/issues?q=is%3Aclosed) issue matching what I'm seeing

### Setup

 - Which version of Git for Windows are you using? Is it 32-bit or 64-bit?

2.13.0 64bit and 2.16.1.4 64bit

```
$ git --version --build-options
git version 2.13.0.windows.1
built from commit: eba7af3dbb4c846c6303c5f64102acee696c9ab0
sizeof-long: 4
machine: x86_64
```

 - Which version of Windows are you running? Vista, 7, 8, 10? Is it 32-bit or 64-bit?

```
$ cmd.exe /c ver

Microsoft Windows [Version 10.0.15063]
```

 - What options did you set as part of the installation? Or did you choose the
   defaults?

```
# One of the following:
> type "C:\Program Files\Git\etc\install-options.txt"
> type "C:\Program Files (x86)\Git\etc\install-options.txt"
> type "%USERPROFILE%\AppData\Local\Programs\Git\etc\install-options.txt"
$ cat /etc/install-options.txt

Does not apply: We are running from a shared network installation.
```

 - Any other interesting things about your environment that might be related
   to the issue you're seeing?

We have recently migrated our data over a CIFS share from on a NetApp FAS2620 running Cluster Mode Release 9.2.

### Details

 - Which terminal/shell are you running Git from? e.g Bash/CMD/PowerShell/other

Bash and CMD

 - What commands did you run to trigger this issue? If you can provide a
   [Minimal, Complete, and Verifiable example](http://stackoverflow.com/help/mcve)
   this will help us understand the issue.

```
thomas@MORGOTH MINGW64 /T/Level_0
$ find .
.
./Level_1a
./Level_1a/Level_2
./Level_1a/Level_2/Level_3
./Level_1a/Level_2/Level_3/Level_4
./Level_1a/Level_2/Level_3/Level_4/Level_5
./Level_1a/Level_2/Level_3/Level_4/Level_5/john.txt
./Level_1b
./Level_1b/Level_2
./Level_1b/Level_2/Level_3
./Level_1b/Level_2/Level_3/Level_4
./Level_1b/Level_2/Level_3/Level_4/Level_5
./Level_1b/Level_2/Level_3/Level_4/Level_5/Level_6
./Level_1b/Level_2/Level_3/Level_4/Level_5/Level_6/john.txt
find: failed to read file names from file system at or below '.': No such file or directory
```
 - What did you expect to occur after running these commands?

`find` should not raise any errors.

 - What actually happened instead?

`find` errors and stops:

> find: failed to read file names from file system at or below '.': No such file or directory

We also have Cygwin deployed similarly and shared from the network and its `find` command is successful:

```
thomas@MORGOTH /cygdrive/t/Level_0
$ find .
.
./Level_1a
./Level_1a/Level_2
./Level_1a/Level_2/Level_3
./Level_1a/Level_2/Level_3/Level_4
./Level_1a/Level_2/Level_3/Level_4/Level_5
./Level_1a/Level_2/Level_3/Level_4/Level_5/john.txt
./Level_1b
./Level_1b/Level_2
./Level_1b/Level_2/Level_3
./Level_1b/Level_2/Level_3/Level_4
./Level_1b/Level_2/Level_3/Level_4/Level_5
./Level_1b/Level_2/Level_3/Level_4/Level_5/Level_6
./Level_1b/Level_2/Level_3/Level_4/Level_5/Level_6/john.txt
./Level_1c
./Level_1c/Level_2
./Level_1c/Level_2/Level_3
./Level_1c/Level_2/Level_3/Level_4
./Level_1c/Level_2/Level_3/Level_4/Level_5
./Level_1c/Level_2/Level_3/Level_4/Level_5/Level_6
./Level_1c/Level_2/Level_3/Level_4/Level_5/Level_6/Level_7
./Level_1c/Level_2/Level_3/Level_4/Level_5/Level_6/Level_7/john.txt
```

Something we noticed after hours of testing around the problem is that it seems to be related to nesting directories level, anything above 5 will cause git bash `find` to error.

**Comment by dscho** [comment]
Does this also happen with a plain [MSYS2](https://msys2.github.io/)?

**Comment by KelSolaar** [comment]
Hi @dscho,

Sorry for the late answer, I confirm that `find` does work properly with http://repo.msys2.org/distrib/x86_64/msys2-x86_64-20161025.exe

Cheers,

Thomas

**Comment by dscho** [comment]
@KelSolaar what do `pacman -Q findutils` and `which find` say? Both in MSYS2 and Git for Windows (in Git for Windows, `pacman` won't work, but there will be a line in your `/etc/package-versions.txt`)?

**Comment by dscho** [comment]
@KelSolaar can you clone a Git for Windows SDK (`git clone --depth 1 https://github.com/git-for-windows/git-sdk-64`) and see whether `find` works as expected in the `git-bash.exe` started from that clone?

**Comment by c-klapaucius** [comment]
This is related to hash calculation of inode numbers. Trailing slash of directory path should not be included in hash. It was already fixed back in 1997 so I guess that this is a regression.

I've reported the issue [here](https://github.com/Alexpux/Cygwin/issues/37)

**Comment by dscho** [comment]
@c-klapaucius could you identify the fix from 1997? Like, a diff, or a revision?

**Comment by c-klapaucius** [comment]
@dscho Probably it is Cygwin specific fix. Cygwin is case sensitive so the fix may not be applicable to MSYS.
For more info you could open the  "cygwin/winsup/CVSChangeLogs.old/cygwin/ChangeLog-1997" file of MSYS2 runtime sources and search for "Ignore trailing backslash".
You can reproduce the error by using the find command to search in a FAT formatted volume. It will not fail if there are just a few files and directories - there should be actual directory structure on the volume.
You could also easily reproduce by executing `find /d/path -printf "%p %i\n"` and then `find /d/path/ -printf "%p %i\n"` on a FAT volume - inode numbers of /d/path and /d/path/ will be different but they shouldn't.

**Comment by dscho** [comment]
Oh, nevermind, I took the advice in a different ticket to create a RAM disk (`imdisk -a -s 64M -o rem -m Y: -p "/FS:FAT /Q /Y"`) and could reproduce the issue...

Now, to find the time to figure out what is going wrong, and where...

**Comment by c-klapaucius** [comment]
This does the trick:

$ mkdir -p /y/a/b/c/d/e/f && touch /y/a/b/c/d/e/f/file

$ find /y
/y
/y/a
/y/a/b
/y/a/b/c
/y/a/b/c/d
/y/a/b/c/d/e
/y/a/b/c/d/e/f
/y/a/b/c/d/e/f/file
find: failed to read file names from file system at or below '/y': No such file or directory

By the way, the problem does not occur with older git 2.x for windows - I am unsure on the exact version though. It may happen on recent versions due to modifications in MSYS2 runtime or findutils.

**Comment by dscho** [comment]
> It may happen on recent versions due to modifications in MSYS2 runtime or findutils.

That is what I suspect.

**Comment by dscho** [comment]
For the record, I am currently trying to rebuild `findutils` (due `./configure` spawning tons and tons of shell processes, my live antivirus is going wild, so it will take a while) with debug symbols so that I can single-step into the offending `fts_read()` function.

Do not hold your breath for it, I *hope* to be able to continue with this side track tomorrow.

**Comment by c-klapaucius** [comment]
You may take a look into fts.c - search the following line:
`__set_errno (ENOENT);           /* disinformation */`
While there may also be a problem in findutils, I believe that the MSYS2 runtime should be fixed.

**Comment by dscho** [comment]
The problem is in https://github.com/Alexpux/Cygwin/commit/8eae854a915db38d88ad30017537fb0c36e66d51#diff-47ffd31e8cd7e0c954ad3ff6c18d1f29 (which we inherited as git-for-windows/msys2-runtime@34f1f2ed67685c8223db6dcfd7edd253424e55df): the slash that was stripped (correctly) is all of a sudden appended (incorrectly), as the `if (need_directory)` block was moved outside the `if (opt & PC_NOFULL)` block.

@Alexpux I believe this is highly incorrect a change. Do you even remember why you did that? The commit message suggests that the second hunk of the diff was included by mistake.

And the commit message as well as the code comment are *highly* insufficient to defend that change. Please explain.

**Comment by dscho** [comment]
I have a patched `msys-2.0.dll` that should fix the problem: @KelSolaar could you please download
[msys2-runtime-fat-fix.zip](https://github.com/git-for-windows/git/files/1811964/msys2-runtime-fat-fix.zip) and unpack the `msys-2.0.dll` contained therein into your Git's `usr\bin` folder (overwriting the existing file), then test whether it really fixes your problem?

**Comment by Alexpux** [comment]
@dscho as I remember this patch was created by @mingwandroid. Need to review this change again

**Comment by dscho** [comment]
@KelSolaar I already got positive feedback about this change in https://github.com/git-for-windows/git/issues/1536, so if you don't respond, I will take it that the bug is fixed for you, too.

**Comment by dscho** [comment]
> as I remember this patch was created by @mingwandroid. Need to review this change again

@Alexpux the commit has you as committer *and* author. If it really was authored by @mingwandroid, you will want to fix this (should have your committership, but @mingwandroid's authorship).

@mingwandroid does @Alexpux' recollection match yours? Did you write that patch? If so, do you have any record why the conditional block was moved outside the `if (opt & PC_NOFULL)` section (which looks really wrong after staring at the code for half an hour)?

**Comment by dscho** [comment]
@mingwandroid just so you know what partial revert I am talking about: https://github.com/dscho/msys2-runtime/commit/fix-path-conv-with-trailing-slash

**Comment by KelSolaar** [comment]
@dscho : I just tested and confirm this is fixing the issue!

Any timeframe to have that patch included in a full release? I might directly patch our installation.

Thanks a lot,

**Comment by dscho** [comment]
> Any timeframe to have that patch included in a full release?

No, not yet. But the next [snapshot](https://wingit.blob.core.windows.net/files/index.html) will have the fix, and the snapshot versions are as close as I can get to official without a real "official" version. I try to keep `master` as functional as possible, so the difference in robustness to the next full version should be unnoticeable.

**Comment by dscho** [comment]
For the record, this fix is part of v2.16.3.

**Comment by KelSolaar** [comment]
Thanks @dscho! I'll update our version today.
